yjit/src/cruby.rs

```diff
@@ -568,6 +569,53 @@ pub fn rust_str_to_sym(str: &str) -> VALUE
     }
 }
 
+/// A location in Rust code for integrating with debugging facilities defined in C.
+/// Use the [src_loc!] macro to crate an instance.
+pub struct SourceLocation {
+    pub file: CString,
+    pub line: c_int,
+}
+
+/// Make a [SourceLocation] at the current spot.
+macro_rules! src_loc {
+    () => {
+        crate::cruby::SourceLocation{
+            file: std::ffi::CString::new(file!()).unwrap(), // ASCII source file paths
+            line: line!().try_into().unwrap(), // not that many lines
+        }
+    }
+}
+
+pub(crate) use src_loc;
+
+/// Acquire the VM lock, make sure all other Ruby threads are asleep then run
+/// some code while holding the lock. Returns whatever `func` returns.
+/// Use with [src_loc!].
+///
+/// Required for code patching in the presence of ractors.
+pub fn with_vm_lock<F, R>(loc: SourceLocation, func: F) -> R
+    where F: FnOnce() -> R + UnwindSafe
+{
+    let file = loc.file.as_ptr();
+    let line = loc.line;
+
+    unsafe { rb_yjit_vm_lock_then_barrier(file, line) };
+
+    let ret = match catch_unwind(func) {
+        Ok(result) => result,
+        Err(_) => {
+            // Theoretically we can recover from some of these panics,
+            // but it's too late if the unwind reaches here.
+            eprintln!("YJIT panicked while compiling. Aborting...");
```

Comment:
That error message needs improvement. We take the lock for things that aren't compiling.

Maybe, "YJIT panicked while VM lock taken" and then print the source location?

Reply:
Updated at in [68677c8](https://github.com/Shopify/ruby/pull/227/commits/68677c84738bac9248e20fef781b5f9dbca5ccdb).